snitun/client/connector.py

```diff
@@ -6,32 +6,187 @@
 from contextlib import suppress
 import ipaddress
 import logging
-from typing import Any
+from typing import Any, Callable
+from asyncio import Transport, BufferedProtocol
+from aiohttp.web import RequestHandler
 
+from ssl import SSLContext, SSLError
 from ..exceptions import MultiplexerTransportClose, MultiplexerTransportError
 from ..multiplexer.channel import MultiplexerChannel
 from ..multiplexer.core import Multiplexer
 
 _LOGGER = logging.getLogger(__name__)
 
 
+class ChannelTransport(Transport):
+    """An asyncio.Transport implementation for multiplexer channel."""
+
+    _start_tls_compatible = True
+
+    def __init__(
+        self, loop: asyncio.AbstractEventLoop, channel: MultiplexerChannel
+    ) -> None:
+        """Initialize ChannelTransport."""
+        self._channel = channel
+        self._loop = loop
+        self._protocol: asyncio.Protocol | None = None
+        self._pause_future: asyncio.Future[None] | None = None
+        super().__init__(extra={"peername": (str(channel.ip_address), 0)})
+
+    def get_protocol(self) -> asyncio.Protocol:
+        """Return the protocol."""
+        return self._protocol
+
+    def set_protocol(self, protocol: asyncio.Protocol) -> None:
+        """Set the protocol."""
+        if not isinstance(protocol, BufferedProtocol):
+            raise ValueError("Protocol must be a BufferedProtocol")
+        self._protocol = protocol
+
+    def is_closing(self) -> bool:
+        """Return True if the transport is closing or closed."""
+        return self._channel.closing
+
+    def close(self) -> None:
+        """Close the underlying channel."""
+        self._channel.close()
+        self._release_pause_future()
+
+    def write(self, data: bytes) -> None:
+        """Write data to the channel."""
+        if not self._channel.closing:
+            self._channel.write_no_wait(data)
+
+    async def start(self) -> None:
+        """Start reading from the channel.
+
+        As a future improvement, it would be a bit more efficient to
+        have the channel call this as a callback from channel.message_transport.
+        """
+        while True:
+            if self._pause_future:
+                await self._pause_future
+
+            try:
+                from_peer = await self._channel.read()
+            except MultiplexerTransportClose:
+                raise
```

Comment:
Pretty sure connect lost will always get called because close on the finally block where this task is awaited but it should get a test